### PR TITLE
Toponaming: ComplexGeoData dressing

### PR DESCRIPTION
**Pull request by chennes**
Closes https://github.com/FreeCAD/FreeCAD/issues/9116

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9780">pipeline status</a> for feature branch [PR_9780](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9780). Pipeline [903567744](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/903567744) was triggered at [ada00b0](https://github.com/FreeCAD/FreeCAD/pull/9780/commits/ada00b0ec19a6ebf952a9a519d1c7428fd956e6f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by chennes** [comment]
> Uhm tests are failing for elementmap? I have already seen that error, something is not initialized properly
It's probably almost the opposite: I bet we are now double-initializing, I suspect we need to add some protection around that setup code to only run when it hasn't already been done (both here and in ElementMap).

File: src/App/ComplexGeoData.cpp [hunk 1]
```diff
@@ -161,8 +173,235 @@ void ComplexGeoData::getFaces(std::vector<Base::Vector3d> &Points,
...
+        for(auto &elementTypes : getElementTypes()) {
+            if(boost::starts_with(type, elementTypes)) {
+                return type[0];
+            }
+        }
+    }
+    return 0;
+}
+
+MappedName ComplexGeoData::renameDuplicateElement(int index,
```
Comment (Pesc0) [inline src/App/ComplexGeoData.cpp:392]: I think this is already in elementmap.

**Review by Pesc0 — COMMENTED** [review]
Except for the one comment I've left I don't see anything suspicious for now. However the more stuff we move around the more complicated navigating the code gets. I may very well have missed functions copied that shouldn't have / not copied that should have. 

I agree the tests error might be a double initialization, I remember struggling with that. Also the error message is quite cryptic. I had tried debugging but I just descended in a spaghetti mess. We need to find a way to initialize one time only for all tests then. Or clear any initialized context and re-init before each test. Need to study gtest for this.

Also, sorry for the dumb question, but wth is a ComplexGeoData?

**Comment by chennes** [comment]
I found the mistake I made in the refactoring (finally) and corrected it, so the tests should all be passing now. I have not done any meaningful refactoring of the code itself, just got a minimal set of tests of it to establish its current functionality. I did run a clang-format pass, and some clang-tidy work as well. I propose to merge as-is, despite the complexity issues, and open those up for other devs to hack on going forward, now that we've got some tests.

File: tests/src/App/ComplexGeoData.cpp [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    // Assert
+    EXPECT_EQ(cgd().getElementMapSize(), 0);
+}
+
+TEST_F(ComplexGeoDataTest, resetElementMapWithArgument)// NOLINT
+{
+    // Arrange
+    auto elementMap = std::make_shared<Data::ElementMap>();
+    auto mappedName = Data::MappedName(";TestMappedName");
+    auto indexedName = Data::IndexedName(cgd().getElementTypes().front(), 1);
```
Comment (Pesc0) [inline tests/src/App/ComplexGeoData.cpp:368]: if mappedname and indexedname are unimportant (from what I understand they're just to get size 1) I would use more generic names for the sake of simplicity and readability.
- in mappedname that leading ';' might seem an important detail
- in indexedname `cgd().getElementTypes().front()` might suggest some kind of coupling to complexgeodata

Same for a couple of tests below

**Review by Pesc0 — COMMENTED** [review]
Why is https://github.com/FreeCAD/FreeCAD/pull/9780/commits/6e43aac28e04d7631aa435bb74bd5346fcf4a509 needed?

File: tests/src/App/ComplexGeoData.cpp [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+    // Act
+    auto result = cgd().getElementName(name);
+
+    // Assert
+    EXPECT_EQ(result.name, mappedName);
+}
+
+TEST_F(ComplexGeoDataTest, getElementNamesWithPrefix)// NOLINT
+{
+    GTEST_SKIP_("Body of method is commented out");
```
[the remark this answers is not in the document]
Comment (chennes) [reply tests/src/App/ComplexGeoData.cpp:219]: It is now: apparently it was used in a couple of deprecated functions. I will remove it entirely.

File: tests/src/App/ComplexGeoData.cpp [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+}
+
+TEST_F(ComplexGeoDataTest, flushElementMap)// NOLINT
+{
+    // Does nothing in the base class
+}
+
+TEST_F(ComplexGeoDataTest, getElementMapReserve)// NOLINT
+{
+    // Does nothing in the base class
```
[the remark this answers is not in the document]
Comment (chennes) [reply tests/src/App/ComplexGeoData.cpp:424]: Yes, this one might just need to be pure virtual (and then drop the test, of course).

File: src/App/ComplexGeoData.cpp [hunk 5]
```diff
@@ -161,8 +173,231 @@ void ComplexGeoData::getFaces(std::vector<Base::Vector3d> &Points,
...
+    }
+    return _elementMap;
+}
+
+void ComplexGeoData::flushElementMap() const
+{
+}
+
+void ComplexGeoData::setElementMap(const std::vector<MappedElement> &map) {
+    _elementMap = std::make_shared<Data::ElementMap>(); // Get rid of the old one, if any, but make
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/App/ComplexGeoData.cpp:316]: Sure -- is there a reason to add that call?

File: tests/src/App/ComplexGeoData.cpp [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+TEST_F(ComplexGeoDataTest, elementTypeCharMappedNameNoPrefix)// NOLINT
+{
+    // Arrange
+    int size {0};
+    Data::MappedName mappedName;
+    Data::IndexedName indexedName;
+    std::tie(indexedName, mappedName) = createMappedName("TestMappedName:;");
+
+    // Act
+    char elementType = cgd().elementType(mappedName.toConstString(0, size));
```
[the remark this answers is not in the document]
Comment (chennes) [reply tests/src/App/ComplexGeoData.cpp:332]: Yeah, this doesn't make an intermediate string.

File: tests/src/App/ComplexGeoData.cpp (hunk 2 above)
Comment (chennes) [reply tests/src/App/ComplexGeoData.cpp:368]: Good suggestions, I will change.

File: tests/src/App/ComplexGeoData.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (chennes) [reply tests/src/App/ComplexGeoData.cpp:424]: I dropped the function entirely, it's never used in the Toponaming branch.

File: src/App/ComplexGeoData.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (chennes) [reply src/App/ComplexGeoData.cpp:299]: All the time, actually! Most of the time to completely remove the map, though.

**Comment by berniev** [comment]
There are a number of tests that have multiple asserts. Where there is only one act this -may- be ok, but asserts on multiple acts are a no-no. A common rule is one assert per test. A test should only fail for one reason. When a test fails, if there are multiple asserts there is no way to immediately know exactly what went wrong, and when you start using the tests you'll soon find discover how annoying that is. Yes, you'll finish up with a lot of tests - so keep em efficient.

Also, it is common for test names to be in the form Ensure_something_returns_a_value_of_0. This helps readability of these long strings. 

It is also better if the test name describes the expected behavioral result, rather than describing what it does. "cat_has_eaten_dog" is preferable to "testCat".

**Comment by berniev** [comment]
How are the execution times for the tests? I see they pull in the entire application/document/etc which is I think a lot of code?

**Comment by Pesc0** [comment]
> When a test fails, if there are multiple asserts there is no way to immediately know exactly what went wrong

No, gtest tells you exactly which assert failed, the names of the variables involved, and what values they had 

> How are the execution times for the tests

As of now, almost instant.

> they pull in the entire application/document/etc 

This is needed because the class `ElementMap` needs a document to save/restore itself. 

Regarding naming: I am definitely not very good at it. If you spot something that could be improved please suggest it or even better submit a PR. Naming is a minor change so I don't see any problems in merging it.

**Comment by berniev** [comment]
> No, gtest tells you exactly which assert failed, the names of the variables involved, and what values they had

You want to look through all that cruft, or just look at a list and know instantly what happened. You might want to search on this subject for better descriptioin than I can provide

> This is needed because

you realise of course that this then is no longer a unit test, but a big integration test, involving a **_lot_** more code besides the unit under test, right?

As an aside here's a flamegraph of a profile run I did a while ago of some tests which show what happens:
flamegraph

**Comment by berniev** [comment]
> as of now, almost instant.
But quantifiable?

**Comment by Pesc0** [comment]
> you realise of course that this then is no longer a unit test, but a big integration test, involving a lot more code besides the unit under test, right?

Alright, what should we do about it?

> But quantifiable?

```
[----------] Global test environment tear-down
[==========] 352 tests from 31 test suites ran. (59 ms total)
[  PASSED  ] 352 tests.
```

**Comment by berniev** [comment]
Thanks, but any specific data re the current tests, as opposed to all tests? Even before theis new stuff I have been seeing 100ms+ overall.

This is the ongoing tension between unit tests, which by definition test a single unit and are lightning fast, and integration tests which test a bunch of stuff and hence take longer.

Unit tests, again by definition, typically exclude (or mock etc) all other (slow) code except the unit under test.

Imagine that as the testing regime grows, based on somewhere between 5000 and 10000 classes in FreeCAD, there might be eg 50000 tests. To be useful in a TDD environment they must complete quickly. Most of the current unit tests register < 0mS, but some pop up at eg 100 ms. Currently there are only hundreds of tests. Extrapolate this to ... There is MUCH information on the net.

The danger is that the tests take so long to run that they are NOT run. If you are working on a new piece of code it is useful to run tests VERY often, so you know straight away when a change has broken something else. This is why unit tests must run at lightning speed. Once you have experienced this way of working I can assure you you'll never go back.

I have no objection per se to integration tests. They have their place, but I keep arguing for them to be separate as they serve a different purpose which is assuring various units work well together.

This may not be a big impact now, but will become increasingly important over time.

In terms of how to address this, assuming the tests run slower than ideal, again there is much literature. This is not throretical clap trap, but industry standard.

**Comment by chennes** [comment]
The only test that takes a measurable amount of time as far as CLion is concerned is whichever one loads the Xerces-C shared library first (so it depends on whether you are running the whole suite, or just the part you're working on). As of this moment compiling the tests dwarfs running them. This is true whether you are only running the tests for the code you are actively editing, or if you decide to run the entire suite. I've been running the ComplexGeoData test fixture after every change, and have not found it particularly onerous.

In Modern C++ Programming with Test-Driven Development Jeff Langr suggests two things that diverge from the discussion above: first, that in C++ we end up doing far less mocking than in other languages (due to the design of C++, which makes some types of injection that are trivial in, say, Python, very hard if not impossible). Second, he disagrees with the old "one assertion per test" rule. In this case I defer to our actual local expert, @JohnAD, who has vastly more (and more recent) experience in such matters and has been mentoring @Pesc0 and I through this process. With his help we're (slowly) gaining the needed experience. Only a few tens of thousands of lines more to go!

**Comment by berniev** [comment]
> that in C++ we end up doing far less mocking

Sure you are not confusing Langer's comment on p129 that in C++ there is far less use of _Direct Injection tools_? As C++ does not have reflection it is indeed very difficult to build direct inject injection tools. Direct Injection tools instantiate objects with required constructor or setter dependencies.

**Comment by berniev** [comment]
Langer page 180:

```
Should you ever have more than one assert per test? Strive hard to have only one. 
Sometimes it makes sense to have more though.

Assertions are postconditions. If multiple assertions are required to describe a single
 behaviour, you can justify a second assertion in a test.

```

Clearly you have misquoted him.

**Comment by berniev** [comment]
> compiling the tests dwarfs running them

You've mentioned this before so I'm curious.

Its true in the obvious sense that a test runs sub-mS, compile seconds, so that's not noteworthy.

So in what sense is it a problem?

I get that the **first** compile takes a while as it sorts out everything involved, but after that if only the test code has changed it should be fast (depending on what the change impacts). On mine its typically just a few seconds. It makes no difference whether running just tests under edit or the whole suite.

Are you experiencing something different? Do you use ninja/ccache?

**Comment by berniev** [comment]
Thinking a bit more about DI / mocks in C++ perhaps you were alluding to C++ not having automatically generated tests and mocks etc like some other languages do. As a result, yes, it could be said mocking is likely not as prevalent in C++ is statistically correct. But it is falacial argument to suggest this is a reason not to mock in C++.

**Pull request merged**